### PR TITLE
Replace copy_file/copy_tree with install/install_tree separately for packages: cub, fastqc, kaldi, sctk.

### PR DESCRIPTION
**Pull request by weijianwen**
Replace `distutils.dir_util.copy_tree` and `copy_file` with Spack's built-in `install_tree` and `install` separately.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Did you test these changes? I think I had a problem running `install_tree('.', prefix)` because the installation directory already existed. Of course, we could patch `install_tree` to work properly in this case. Or just use `copy_tree`.

**Comment by weijianwen** [comment]
I tested in `cub` in which `install_tree` works well. Shall I test more packages?

**Comment by adamjstewart** [comment]
That's because `prefix.include` doesn't exist yet. Try one that installs to `prefix`.

**Comment by weijianwen** [comment]
I'll try one and update soon.

**Comment by hartzell** [comment]
Looking through the files changed, the only ones that I'm familiar with are *fastqc* and *jdk*.

I tried to install *fastqc*, and it failed like so (as @adamjstewart predicted...):

```
[hartzelg@lb046dev spack-install-tree-branch]$ spack install fastqc
==> Installing fastqc
==> Installing jdk
==> Fetching http://download.oracle.com/otn-pub/java/jdk/8u92-b14/jdk-8u92-linux-x64.tar.gz
######################################################################## 100.0%
==> Staging archive: /home/hartzelg/tmp/spack-install-tree-branch/var/spack/stage/jdk-8u92-vb2ldslwgdsfbozyx727qhkhxnrlionr/jdk-8u92-linux-x64.tar.gz
==> Created stage in /home/hartzelg/tmp/spack-install-tree-branch/var/spack/stage/jdk-8u92-vb2ldslwgdsfbozyx727qhkhxnrlionr
==> Ran patch() for jdk
==> Building jdk [Package]
==> Executing phase : 'install'
==> Error: OSError: [Errno 17] File exists: '/home/hartzelg/tmp/spack-install-tree-branch/opt/spack/linux-centos7-x86_64/gcc-4.8.5/jdk-8u92-vb2ldslwgdsfbozyx727qhkhxnrlionr'
/home/hartzelg/tmp/spack-install-tree-branch/var/spack/repos/builtin/packages/jdk/package.py:55, in install:
     54       def install(self, spec, prefix):
  >> 55           install_tree(".", prefix)

See build log for details:
  /tmp/hartzelg/spack-stage/spack-stage-m7Iv8I/jdk1.8.0_92/spack-build.out
[hartzelg@lb046dev spack-install-tree-branch]$ tail /tmp/hartzelg/spack-stage/spack-stage-m7Iv8I/jdk1.8.0_92/spack-build.out
==> Installing . to /home/hartzelg/tmp/spack-install-tree-branch/opt/spack/linux-centos7-x86_64/gcc-4.8.5/jdk-8u92-vb2ldslwgdsfbozyx727qhkhxnrlionr
```

**Comment by weijianwen** [comment]
I traced back to `lib.spack.util.filesystem`, finding that `install_tree` is implemented with `shutil.copytree(str, dst)` which requires `dst` must NOT exist. So operation `install(., prefix)` fails. 

Luckily, only `gradle` and `jdk` packages fits into this situation. I'll test and revert failing packages soon.

**Comment by weijianwen** [comment]
I've tested cub, fastqc, kaldi and sctk. This commit works. It is ready for merge now.

**Comment by adamjstewart** [comment]
> I traced back to `lib.spack.util.filesystem`, finding that `install_tree` is implemented with `shutil.copytree(str, dst)` which requires dst must NOT exist. So operation `install_tree('.', prefix)` fails.

Why not implement it with `distutils.dir_util.copy_tree` instead? Does anyone know of any important differences that we would care about (perhaps differences in how it treats permissions)?

File: var/spack/repos/builtin/packages/fastqc/package.py [hunk 1]
```diff
@@ -41,15 +39,14 @@ class Fastqc(Package):
...
+        install('fastqc', self.prefix.bin)
+        set_executable(join_path(self.prefix.bin, 'fastqc'))
         for j in ['cisd-jhdf5.jar', 'jbzip2-0.9.jar', 'sam-1.103.jar']:
-            copy_file(j, self.prefix.lib)
+            install(j, join_path(self.prefix.lib, j))
         for d in ['Configuration', 'net', 'org', 'Templates', 'uk']:
-            copy_tree(d, join_path(self.prefix.lib, d))
-        chmod = which('chmod')
-        chmod('+x', join_path(self.prefix.bin, 'fastqc'))
+             install_tree(d, join_path(self.prefix.lib, d))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fastqc/package.py]: This line is indented one space too far and is causing the flake8 tests to fail.

**Comment by becker33** [comment]
@hartzell, does this now satisfy your review?

@adamjstewart I don't know of any reason not to use the `distutils.dir_util.copy_tree`. The only concern would be if there are build systems out there for which we would be clobbering files somehow, but I can't imagine how that would occur.

**Comment by hartzell** [comment]
Apologies for the delay, I'm still getting used to being asked for reviews....

I cloned this repo, checked out the branch and tried to `spack install jdk`.  It failed as follows:

```
hartzelg@blah:hack (install-tree)$ (module purge;./bin/spack install jdk)
Traceback (most recent call last):
  File "./bin/spack", line 220, in <module>
    main(sys.argv)
  File "./bin/spack", line 216, in main
    _main(args, unknown)
  File "./bin/spack", line 183, in _main
    return_val = command(parser, args)
  File "/home/hartzelg/tmp/hack/lib/spack/spack/cmd/install.py", line 317, in install
    specs = spack.cmd.parse_specs(args.package, concretize=True)
  File "/home/hartzelg/tmp/hack/lib/spack/spack/cmd/__init__.py", line 116, in parse_specs
    spec.concretize()  # implies normalize
  File "/home/hartzelg/tmp/hack/lib/spack/spack/spec.py", line 1681, in concretize
    self._expand_virtual_packages(),
  File "/home/hartzelg/tmp/hack/lib/spack/spack/spec.py", line 1598, in _expand_virtual_packages
    spec)
  File "/home/hartzelg/tmp/hack/lib/spack/spack/concretize.py", line 113, in choose_virtual_or_external
    candidates = self._valid_virtuals_and_externals(spec)
  File "/home/hartzelg/tmp/hack/lib/spack/spack/concretize.py", line 87, in _valid_virtuals_and_externals
    if is_spec_buildable(cspec):
  File "/home/hartzelg/tmp/hack/lib/spack/spack/package_prefs.py", line 239, in is_spec_buildable
    allpkgs = get_packages_config()
  File "/home/hartzelg/tmp/hack/lib/spack/spack/package_prefs.py", line 53, in get_packages_config
    if spack.repo.is_virtual(pkg_name)]
  File "/home/hartzelg/tmp/hack/lib/spack/spack/repository.py", line 347, in is_virtual
    return pkg_name in self.provider_index
  File "/home/hartzelg/tmp/hack/lib/spack/spack/repository.py", line 233, in provider_index
    self._provider_index.merge(repo.provider_index)
  File "/home/hartzelg/tmp/hack/lib/spack/spack/repository.py", line 652, in provider_index
    self._update_provider_index()
  File "/home/hartzelg/tmp/hack/lib/spack/spack/repository.py", line 646, in _update_provider_index
    self._provider_index.to_yaml(new)
  File "/home/hartzelg/tmp/hack/lib/spack/llnl/util/lock.py", line 312, in __exit__
    if self._as.__exit__(type, value, traceback):
  File "/home/hartzelg/tmp/hack/lib/spack/spack/file_cache.py", line 152, in __exit__
    raise value
TypeError: unhashable type: 'syaml_list'
```

**Comment by adamjstewart** [comment]
See #8182 for continued work on this.

**Pull request closed without merging**